test/src/unit-cppapi-array.cc

```diff
@@ -1105,3 +1107,72 @@ TEST_CASE("C++ API: Test heterogeneous dimensions", "[cppapi][sparse][heter]") {
   if (vfs.is_dir(array_name))
     vfs.remove_dir(array_name);
 }
+
+TEST_CASE(
+    "C++ API: Test string dimensions, 1d",
+    "[cppapi][sparse][string-dims][1d]") {
+  const std::string array_name = "cpp_unit_array";
+  Context ctx;
+  VFS vfs(ctx);
+
+  if (vfs.is_dir(array_name))
+    vfs.remove_dir(array_name);
+
+  // Create array
+  auto d = Dimension::create(ctx, "d", TILEDB_STRING_ASCII, nullptr, nullptr);
+  Domain dom(ctx);
+  dom.add_dimension(d);
+  ArraySchema schema(ctx, TILEDB_SPARSE);
+  auto a = Attribute::create<int32_t>(ctx, "a");
+  schema.add_attribute(a);
+  schema.set_domain(dom);
+  Array::create(array_name, schema);
+
+  // Write
+  Array array(ctx, array_name, TILEDB_WRITE);
+  std::vector<int32_t> buff_a = {3, 2, 1, 4};
+  std::string d_data("ccbbddddaa");
+  uint64_t d_off[] = {0, 2, 4, 8};
+  uint64_t d_off_size = 4;
+  Query query(ctx, array, TILEDB_WRITE);
+  CHECK_NOTHROW(query.set_buffer(
+      "d", d_off, d_off_size, (void*)d_data.c_str(), d_data.size()));
+  query.set_buffer("a", buff_a);
+  query.set_layout(TILEDB_UNORDERED);
+  CHECK_NOTHROW(query.submit());
+  array.close();
+
+  // Non-empty domain
+  Array array_r(ctx, array_name, TILEDB_READ);
+  auto non_empty_domain = array_r.non_empty_domain_var("d");
+  CHECK(non_empty_domain.first == std::string("aa"));
+  CHECK(non_empty_domain.second == std::string("dddd"));
+  non_empty_domain = array_r.non_empty_domain_var(0);
+  CHECK(non_empty_domain.first == std::string("aa"));
+  CHECK(non_empty_domain.second == std::string("dddd"));
+  CHECK_THROWS(array_r.non_empty_domain_var(1));
+  CHECK_THROWS(array_r.non_empty_domain_var("foo"));
+
+  // Read
+  std::string s1("a", 1);
+  std::string s2("ee", 2);
```

Comment:
Setting `["aa", "bb"]` or `["a", "bb"]` as the range I get back `"[4 2]"`
setting `["aa", "b"]` or `["a", "b"]` as the range I get back `[4]`.

I think this is correct given the comparator implementation, but it seems to mean that inside the nonempty_domain they have to know the longest prefix boundaries a priori in order to index all strings starting with a minimal prefix (e.g. "b" here).

Reply:
Yeah, the query in that case should have been `["a", "b\0"]`, where `\0` has `char` value `0`. Prefix queries (though common) are quite special, so we should probably handle them specially by exposing more explicit API functions.